source/extensions/tracers/xray/tracer.cc

```diff
@@ -1,37 +1,133 @@
 #include "extensions/tracers/xray/tracer.h"
 
+#include <sys/socket.h>
+
 #include <algorithm>
 #include <array>
+#include <chrono>
 #include <string>
 
 #include "envoy/http/header_map.h"
+#include "envoy/network/listener.h"
 
 #include "common/common/fmt.h"
 #include "common/common/hex.h"
+#include "common/runtime/runtime_impl.h"
+
+#include "source/extensions/tracers/xray/daemon.pb.h"
 
 namespace Envoy {
 namespace Extensions {
 namespace Tracers {
 namespace XRay {
 
+namespace {
+constexpr auto XRaySerializationVersion = "1";
+
+// X-Ray Trace ID Format
+//
+// A trace_id consists of three parts separated by hyphens.
+// For example, 1-58406cf0-a006649127e371903a2de979.
+// This includes:
+//
+// - The version number, that is, 1.
+// - The time of the original request, in Unix epoch time, in 8 hexadecimal digits.
+// - A 96-bit unique identifier in 24 hexadecimal digits.
+//
+// For more details see:
+// https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html#api-segmentdocuments-fields
+std::string generateTraceId(Envoy::SystemTime point_in_time) {
+  using namespace std::chrono;
+  Runtime::RandomGeneratorImpl rng;
+  const auto epoch = time_point_cast<seconds>(point_in_time).time_since_epoch().count();
+  std::string out;
+  out.reserve(35);
+  out += XRaySerializationVersion;
+  out.push_back('-');
+  // epoch in seconds represented as 8 hexadecimal characters
+  out += Hex::uint32ToHex(epoch);
+  out.push_back('-');
+  std::string uuid = rng.uuid();
+  // unique id represented as 24 hexadecimal digits and no dashes
+  uuid.erase(std::remove(uuid.begin(), uuid.end(), '-'), uuid.end());
+  ASSERT(uuid.length() >= 24);
+  out += uuid.substr(uuid.length() - 24);
```

Comment:
should this be `uuid.substr(0, 24)` to get 24 characters?

Reply:
`out += uuid.substr(uuid.length() - 24);` I can't remember the reason I was adding the **last** 24 hex characters rather than the first 24.

I'll change it. Thanks